operators-installer - init add

itewk opened the pull request:
#### What is this PR About?
Adds `operators-installer` helm chart which allows for declarative control of managing operator versions.

#### How do we test this?
Target the helm chart against a running OCP cluster.

I suggest testing with the following configs:

First
```yaml
operators:
- channel: stable
  installPlanApproval: Manual
  name: external-secrets-operator
  source: community-operators
  sourceNamespace: openshift-marketplace
  csv: external-secrets-operator.v0.8.2
```

Then:
```yaml
operators:
- channel: stable
  installPlanApproval: Manual
  name: external-secrets-operator
  source: community-operators
  sourceNamespace: openshift-marketplace
  csv: external-secrets-operator.v0.8.3
```

eformat commented:
great stuff @itewk  👍 could you pop an entry into the test harness as well please for the new chart:

https://github.com/redhat-cop/helm-charts/blob/master/_test/conftest.sh

there is a simpler chart here, that i think we could replace with yours:

https://github.com/redhat-cop/helm-charts/tree/master/charts/operatorhub

the only missing thing in your chart is OpertorGroup creation ? which is pretty handy to have ... WDYT about adding it in to your chart as well ?

https://github.com/redhat-cop/helm-charts/blob/master/charts/operatorhub/templates/operatorgroup.yaml

then we could perhaps deprecate the other one

sabre1041 commented on charts/operators-installer/templates/Job_installplan-approver.yml:
```diff
@@ -0,0 +1,90 @@
+    spec:
+      containers:
+      - name: installplan-approver
+        image: registry.redhat.io/openshift4/ose-cli:v4.10
```
Let's go ahead an parameterize the image field

itewk replied:
done.

sabre1041 reviewed (changes requested):
Looks really good. One request

itewk commented:
> great stuff @itewk 👍 could you pop an entry into the test harness as well please for the new chart:
> 
> https://github.com/redhat-cop/helm-charts/blob/master/_test/conftest.sh

done. 

> there is a simpler chart here, that i think we could replace with yours:
> 
> https://github.com/redhat-cop/helm-charts/tree/master/charts/operatorhub
> 
> the only missing thing in your chart is OpertorGroup creation ? which is pretty handy to have ... WDYT about adding it in to your chart as well ?
> 
> https://github.com/redhat-cop/helm-charts/blob/master/charts/operatorhub/templates/operatorgroup.yaml
> 
> then we could perhaps deprecate the other one

done. I simplieifed it to onlhy allow for one namepsace, the namespace the chart is being installed in based on the latest docs (https://docs.openshift.com/container-platform/4.13/operators/understanding/olm/olm-understanding-operatorgroups.html) that say:

>   | Listing multiple namespaces via spec.targetNamespaces or use of a label selector via spec.selector  is not recommended, as the support for more than one target namespace  in an Operator group will likely be removed in a future release.
> -- | --


I tested this on my own cluster installing in both the openshift-operators namespace without creating a group and then in a custom namespace with creating a group.

itewk commented:
@eformat @sabre1041 its also worth mentioning that the idea for this came from https://github.com/redhat-cop/gitops-catalog/blob/main/installplan-approver/base/installplan-approver-job.yaml. This just super charges that idea.

eformat commented:
> I tested this on my own cluster installing in both the openshift-operators namespace without creating a group and then in a custom namespace with creating a group.

I think all that is required is the ability for the chart to create the Operator Group as well if specified for a custom namespace ? Otherwise the chart will fail to install.

eformat commented:
> @eformat @sabre1041 its also worth mentioning that the idea for this came from https://github.com/redhat-cop/gitops-catalog/blob/main/installplan-approver/base/installplan-approver-job.yaml. This just super charges that idea.

ah for sure .. there are charts in this repo that pioneered that idea as well :)

itewk commented:
> > I tested this on my own cluster installing in both the openshift-operators namespace without creating a group and then in a custom namespace with creating a group.
> 
> I think all that is required is the ability for the chart to create the Operator Group as well if specified for a custom namespace ? Otherwise the chart will fail to install its listed operators ?

That’s already in here now. I added the template and a new param and updated the doc all this morning.

eformat commented:
ok .. sorry i missed it .. hold on just checking

eformat commented:
ah yup ! sweet .. LGTM

eformat commented:
cheers Ian .. lets go! 💫
